### ISSUE TITLE
Rendering Contexts vs. Devices ?

### ISSUE THREAD

**Issue by monaghanwashere** [kcat/openal-soft#711]
I want to use OpenAL Soft to render audio streams for multiple listeners. The general idea is that, there is a central server with multiple users connected to it. The server contains all the information about the world that the users are in, and their positions in the world, etc. On the server, I want to use OpenAL Soft to render multiple streams, one for each user, and then the server will send all these uniquely rendered streams back to each client.

Now I know that each OpenAL listener has its own context, with its own sources, their own positions, etc. I also know that I need to use a loopback device on the server to render the streams for each listener, because the rendered audio stream needs to be sent over a network (instead of played out over audio hardware).

My question is, on the server, is creating multiple contexts on a single loopback device possible, or would I need to create multiple loopback devices? The former seems simpler, but my understanding is also that rendering samples for a given device will render ALL contexts for that device, and since I want multiple streams, I will have no choice but to create multiple devices (what confuses me, is that how can you even have multiple listeners per device, but a single render?)

Any insight is appreciated, thanks!

**Comment by kcat** [comment]
> My question is, on the server, is creating multiple contexts on a single loopback device possible, or would I need to create multiple loopback devices? The former seems simpler, but my understanding is also that rendering samples for a given device will render ALL contexts for that device, and since I want multiple streams, I will have no choice but to create multiple devices (what confuses me, is that how can you even have multiple listeners per device, but a single render?)

This is correct, a single device with multiple contexts/listeners will mix all contexts/listeners together for the device output. This would be useful where you want separate contexts on a single device output, like a separate output media filter (for GStreamer or whatever Windows has these days) that uses the same device as the game, a picture-in-picture scenario where you can see and hear from a second viewpoint along with the normal one, split-screen multiplayer, etc.

If you want separate output streams for each context/listener, you need to use a separate device for each.
